File: codeforces/round403/B.cc
// Codeforces Round #403 (Div. 2)
// B. The Meeting Place Cannot Be Changed
// http://codeforces.com/contest/782/problem/B

#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

vector<double> coords, speeds;

bool canMeet(double t){
	double L1 = coords[0] - speeds[0] * t;
	double R1 = coords[0] + speeds[0] * t;

	for(int i=1; i<coords.size(); ++i){
		double L2 = coords[i] - speeds[i] * t;
		double R2 = coords[i] + speeds[i] * t;
		L1 = max(L1, L2);
		R1 = min(R1, R2);
		if(L1 > R1) return false;
	}
	
	return true;
}

int main() {
	int n;
	scanf("%d", &n);

	coords.resize(n);
	speeds.resize(n);
	for(int i=0; i<n; ++i) scanf("%lf", &coords[i]);
	for(int i=0; i<n; ++i) scanf("%lf", &speeds[i]);

	double lb = 0.0, ub = 1e9;
	for(int i=0; i<100; ++i){
		double mid = (lb + ub) / 2;
		if(canMeet(mid)){
			ub = mid;
		}else{
			lb = mid;
		}
	}

	double answer = (lb + ub) / 2;
	printf("%.16lf\n", answer);

	return 0;
}

File: codeforces/round403/B.hs
-- Codeforces Round #403 (Div. 2)
-- B. The Meeting Place Cannot Be Changed
-- http://codeforces.com/contest/782/problem/B

import Control.Monad
import Data.Maybe
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8

getInt = read `fmap` getLine :: IO Int
getInts = (map parseInt . BS8.words) `fmap` BS.getLine :: IO [Int]
parseInt = fst . fromJust . BS8.readInt

main = do
	n <- getInt
	coords <- (map fromIntegral) `fmap` getInts :: IO [Double]
	speeds <- (map fromIntegral) `fmap` getInts :: IO [Double]
	let (lb, ub) = recurse coords speeds (0.0, 1e9) 100
	let answer = (lb + ub) / 2.0
	print answer

recurse _ _ result 0 = result
recurse coords speeds (lb, ub) cnt
	| canMeet coords speeds mid = recurse coords speeds (lb, mid) (cnt-1)
	| otherwise = recurse coords speeds (mid, ub) (cnt-1)
	where mid = (lb + ub) / 2.0

canMeet :: [Double] -> [Double] -> Double -> Bool
canMeet coords speeds t = isJust $ foldl getOverlap (Just (0.0, 1e9)) ranges where
	ranges = zipWith (\coord speed -> (coord - speed * t, coord + speed * t)) coords speeds

getOverlap Nothing _ = Nothing
getOverlap (Just (l1, r1)) (l2, r2)
	| l > r = Nothing
	| otherwise = Just (l, r)
	where
		l = max l1 l2
		r = min r1 r2
